src/BenchmarkDotNet/Engines/EngineWarmupStage.cs

```diff
@@ -1,83 +1,24 @@
-            => Run(invokeCount, IterationMode.Workload, false, unrollFactor, forceSpecific);
+            => Run(invokeCount, IterationMode.Overhead, unrollFactor, RunStrategy.Throughput);
 
-        internal List<Measurement> Run(long invokeCount, IterationMode iterationMode, bool runAuto, int unrollFactor, bool forceSpecific = false)
-            => (runAuto || warmupCount == null || warmupCount.Value == EngineResolver.ForceAutoWarmup) && !forceSpecific
-                ? RunAuto(invokeCount, iterationMode, unrollFactor)
-                : RunSpecific(invokeCount, iterationMode, warmupCount ?? 0, unrollFactor);
-
-        private List<Measurement> RunAuto(long invokeCount, IterationMode iterationMode, int unrollFactor)
-        {
-            var measurements = new List<Measurement>(maxIterationCount);
-            int iterationCounter = 0;
-            while (true)
-            {
-                iterationCounter++;
-                measurements.Add(RunIteration(iterationMode, IterationStage.Warmup, iterationCounter, invokeCount, unrollFactor));
-                if (IsWarmupFinished(measurements, iterationMode))
-                    break;
-            }
-            WriteLine();
-
-            return measurements;
-        }
+        public void RunWorkload(long invokeCount, int unrollFactor, RunStrategy runStrategy)
+            => Run(invokeCount, IterationMode.Workload, unrollFactor, runStrategy);
 
-        private List<Measurement> RunSpecific(long invokeCount, IterationMode iterationMode, int iterationCount, int unrollFactor)
+        internal List<Measurement> Run(long invokeCount, IterationMode iterationMode, int unrollFactor, RunStrategy runStrategy)
         {
-            var measurements = new List<Measurement>(maxIterationCount);
-            for (int i = 0; i < iterationCount; i++)
-                measurements.Add(RunIteration(iterationMode,IterationStage.Warmup, i + 1, invokeCount, unrollFactor));
-
-            WriteLine();
-
-            return measurements;
-        }
-
-        private bool IsWarmupFinished(List<Measurement> measurements, IterationMode iterationMode)
-        {
-            int n = measurements.Count;
-            if (n >= maxIterationCount || iterationMode == IterationMode.Overhead && n >= MaxOverheadIterationCount)
-                return true;
-            if (n < minIterationCount)
-                return false;
-
-            int dir = -1, changeCount = 0;
-            for (int i = 1; i < n; i++)
-            {
-                int nextDir = Math.Sign(measurements[i].Nanoseconds - measurements[i - 1].Nanoseconds);
-                if (nextDir != dir || nextDir == 0)
-                {
-                    dir = nextDir;
-                    changeCount++;
-                }
-            }
-
-            return changeCount >= 4;
+            var criteria = DefaultStoppingCriteriaFactory.Instance.CreateWarmup(engine.TargetJob, engine.Resolver, iterationMode, runStrategy);
```

Comment:
how can be the criteria customized? `DefaultStoppingCriteriaFactory.Instance` always points to a `DefaultStoppingCriteriaFactory` and is a read-only field

Reply:
I will add API for customization in the future. It will be pretty hard to implement the whole refactoring at once. In this PR, I just introduce `IStoppingCriteria` as a concept.